File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        deque<int> maxDeque, minDeque;
        int left = 0, right;
        int maxLength = 0;

        for (right = 0; right < nums.size(); ++right) {
            // Maintain the decreasing deque for the max value in the current window
            while (!maxDeque.empty() && nums[maxDeque.back()] <= nums[right])
                maxDeque.pop_back();
            maxDeque.push_back(right);

            // Maintain the increasing deque for the min value in the current window
            while (!minDeque.empty() && nums[minDeque.back()] >= nums[right])
                minDeque.pop_back();
            minDeque.push_back(right);

            // If the window is invalid, move the left pointer to the right
            while (nums[maxDeque.front()] - nums[minDeque.front()] > limit) {
                ++left;
                if (maxDeque.front() < left) maxDeque.pop_front();
                if (minDeque.front() < left) minDeque.pop_front();
            }

            // Update the maxLength
            maxLength = max(maxLength, right - left + 1);
        }

        return maxLength;
    }
};

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cs
public class Solution {
    public int LongestSubarray(int[] nums, int limit) {
        Deque<int> maxDeque = new Deque<int>();
        Deque<int> minDeque = new Deque<int>();
        int left = 0, right;
        int maxLength = 0;

        for (right = 0; right < nums.Length; ++right) {
            // Maintain the decreasing deque for the max value in the current window
            while (maxDeque.Count > 0 && nums[maxDeque.Back] <= nums[right])
                maxDeque.PopBack();
            maxDeque.PushBack(right);

            // Maintain the increasing deque for the min value in the current window
            while (minDeque.Count > 0 && nums[minDeque.Back] >= nums[right])
                minDeque.PopBack();
            minDeque.PushBack(right);

            // If the window is invalid, move the left pointer to the right
            while (nums[maxDeque.Front] - nums[minDeque.Front] > limit) {
                ++left;
                if (maxDeque.Front < left) maxDeque.PopFront();
                if (minDeque.Front < left) minDeque.PopFront();
            }

            // Update the maxLength
            maxLength = Math.Max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

public class Deque<T> {
    private LinkedList<T> list = new LinkedList<T>();

    public void PushBack(T value) {
        list.AddLast(value);
    }

    public void PopBack() {
        if (list.Count > 0)
            list.RemoveLast();
    }

    public T Back {
        get { return list.Last.Value; }
    }

    public void PushFront(T value) {
        list.AddFirst(value);
    }

    public void PopFront() {
        if (list.Count > 0)
            list.RemoveFirst();
    }

    public T Front {
        get { return list.First.Value; }
    }

    public int Count {
        get { return list.Count; }
    }
}